interfaces/tests/file_int/file_int.c

```diff
@@ -0,0 +1,427 @@
+/*Tests_SRS_FILE_43_016:[file_read_async shall call user_callback passing user_context and success depending on the success of the asynchronous read operation.]*/
+/*Tests_SRS_FILE_43_031 : [file_read_async shall succeed and return FILE_READ_ASYNC_OK.]*/
+TEST_FUNCTION(write_twice_to_a_file_contiguously_and_read_from_it)
+{
+    ///arrange
+    const int size = 5;
+
+    unsigned char source1[] = "abcd";
+    WRITE_COMPLETE_CONTEXT write_context1;
+    write_context1.pre_callback_value = 41;
+    interlocked_exchange(&write_context1.value, write_context1.pre_callback_value);
+    write_context1.post_callback_value = 42;
+
+    unsigned char source2[] = "efgh";
+    WRITE_COMPLETE_CONTEXT write_context2;
+    write_context2.pre_callback_value = 45;
+    interlocked_exchange(&write_context2.value, write_context2.pre_callback_value);
+    write_context2.post_callback_value = 46;
+
+    unsigned char destination[9];
+    READ_COMPLETE_CONTEXT read_context;
+    read_context.pre_callback_value = 43;
+    interlocked_exchange(&read_context.value, read_context.pre_callback_value);
+    read_context.post_callback_value = 44;
+
+
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(NULL);
+    FILE_HANDLE file_handle = file_create(execution_engine, "write_twice_to_a_file_contiguously_and_read_from_it.txt", NULL, NULL);
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT write_result1 = file_write_async(file_handle, source1, size, 0, write_callback, &write_context1);
+    wait_on_address(&write_context1.value, &write_context1.pre_callback_value, UINT32_MAX);
+    FILE_WRITE_ASYNC_RESULT write_result2 = file_write_async(file_handle, source2, size, 4, write_callback, &write_context2);
+    wait_on_address(&write_context2.value, &write_context2.pre_callback_value, UINT32_MAX);
+    FILE_READ_ASYNC_RESULT read_result = file_read_async(file_handle, destination, 9, 0, read_callback, &read_context);
+    wait_on_address(&read_context.value, &read_context.pre_callback_value, UINT32_MAX);
+
+    ///assert
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_OK, write_result1);
+    ASSERT_ARE_EQUAL(int32_t, write_context1.post_callback_value, interlocked_or(&write_context1.value, 0), "value should be post_callback_value");
+    ASSERT_IS_TRUE(write_context1.did_write_succeed);
+
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_OK, write_result2);
+    ASSERT_ARE_EQUAL(int32_t, write_context2.post_callback_value, interlocked_or(&write_context2.value, 0), "value should be post_callback_value");
+    ASSERT_IS_TRUE(write_context2.did_write_succeed);
+
+    ASSERT_ARE_EQUAL(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_OK, read_result);
+    ASSERT_ARE_EQUAL(int32_t, read_context.post_callback_value, interlocked_or(&read_context.value, 0), "value should be post_callback_value");
+    ASSERT_IS_TRUE(read_context.did_read_succeed);
+
+    ASSERT_ARE_EQUAL(char_ptr, "abcdefgh", destination);

```

Comment:
>ASSERT_ARE_EQUAL(char_ptr, "abcdefgh", destination); [](start = 4, length = 52)

Is this working? It looks like you write: "abcd\0efgh\0" because you are writing 5 bytes each time...

Reply:
yeah because the first \0 gets overwritten

---
In reply to: [456715740](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456715740) [](ancestors = 456715740)